File: stdlib/public/LSP/IndexStoreDB/IndexStoreDB_LLVMSupport/Options.cpp
//===- toolchain/Support/Options.cpp - Debug options support ---------*- C++ -*-===//
//
// Copyright (c) 2025, NeXTHub Corporation. All Rights Reserved.
// DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
// 
// Author: Tunjay Akbarli
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at:
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// 
// Please contact NeXTHub Corporation, 651 N Broad St, Suite 201,
// Middletown, DE 19709, New Castle County, USA.
//
//===----------------------------------------------------------------------===//
//
// This file implements the helper objects for defining debug options using the
// new API built on cl::opt, but not requiring the use of static globals.
//
//===----------------------------------------------------------------------===//

#include <IndexStoreDB_LLVMSupport/toolchain_Support_Options.h>
#include <IndexStoreDB_LLVMSupport/toolchain_Support_ManagedStatic.h>

using namespace toolchain;

OptionRegistry::~OptionRegistry() {
  for (auto IT = Options.begin(); IT != Options.end(); ++IT)
    delete IT->second;
}

void OptionRegistry::addOption(void *Key, cl::Option *O) {
  assert(Options.find(Key) == Options.end() &&
         "Argument with this key already registerd");
  Options.insert(std::make_pair(Key, O));
}

static ManagedStatic<OptionRegistry> OR;

OptionRegistry &OptionRegistry::instance() { return *OR; }

File: compiler/internal/vm/static_core/plugins/ets/playground/frontend/src/store/slices/options.ts
/*
 * Copyright (c) NeXTHub Corporation. All Rights Reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * Author: Tunjay Akbarli
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at:
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Please contact NeXTHub Corporation, 651 N Broad St, Suite 201,
 * Middletown, DE 19709, New Castle County, USA.
 */

import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { IOption } from '../../models/options';

export interface IObj {
    [key: string]: boolean | number | string;
}

interface IState {
    isLoading: boolean
    options: IOption[] | null
    pickedOptions: IObj | null
}

const initialState: IState = {
    isLoading: false,
    options: null,
    pickedOptions: {},
};

const optionsSlice = createSlice({
    name: 'optionsState',
    initialState,
    reducers: {
        setOptionsLoading(state, action: PayloadAction<boolean>) {
            state.isLoading = action.payload;
        },
        setOptionsResponse(state, action: PayloadAction<IOption[]>) {
            state.options = action.payload;
        },
        setOptionsPicked(state, action: PayloadAction<IObj>) {
            state.pickedOptions = action.payload;
        }
    }
});

export const {
    setOptionsLoading,
    setOptionsResponse,
    setOptionsPicked,
} = optionsSlice.actions;

export default optionsSlice.reducer;
